art/attacks/evasion/wasserstein.py

```diff
@@ -0,0 +1,1151 @@
+            "conjugate_sinkhorn_max_iter": conjugate_sinkhorn_max_iter,
+            "projected_sinkhorn_max_iter": projected_sinkhorn_max_iter,
+            "batch_size": batch_size,
+        }
+
+        Wasserstein.set_params(self, **kwargs)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)` or indices of shape
+                  (nb_samples,). Only provide this parameter if you'd like to use true labels when crafting adversarial
+                  samples. Otherwise, model predictions are used as labels to avoid the "label leaking" effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236). Default is `None`.
+        :type y: `np.ndarray`
+        :param cost_matrix: A non-negative cost matrix.
+        :type cost_matrix: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        y = check_and_transform_label_format(y, self.estimator.nb_classes)
+        x_adv = x.copy().astype(ART_NUMPY_DTYPE)
+
+        if y is None:
+            # Throw error if attack is targeted, but no targets are provided
+            if self.targeted:
+                raise ValueError("Target labels `y` need to be provided for a targeted attack.")
+
+            # Use model predictions as correct outputs
+            targets = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
+        else:
+            targets = y
+
+        # Compute the cost matrix if needed
+        cost_matrix = kwargs.get("cost_matrix")
+        if cost_matrix is None:
+            cost_matrix = self._compute_cost_matrix(self.p, self.kernel_size)
+
+        # Compute perturbation with implicit batching
+        nb_batches = int(np.ceil(x.shape[0] / float(self.batch_size)))
+        for batch_id in range(nb_batches):
+            logger.debug("Processing batch %i out of %i", batch_id, nb_batches)
+
+            batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+            batch = x_adv[batch_index_1: batch_index_2]
+            batch_labels = targets[batch_index_1: batch_index_2]
+
+            x_adv[batch_index_1:batch_index_2] = self._generate_batch(
+                batch,
+                batch_labels,
+                cost_matrix,
+                self.kernel_size,
+                self.max_iter,
+                self.norm,
+                self.ball,
+                self.targeted,
+                self.eps,
+                self.eps_iter,
+                self.eps_factor,
+                self.eps_step,
+                self.regularization,
+                self.conjugate_sinkhorn_max_iter,
+                self.projected_sinkhorn_max_iter,
```

Comment:
Are these arguments required? Could they be replaced by direct access to self inside `self._generate_batch`?

Reply:
Hi @beat-buesser , I suggest to use it this way, we should avoid to access to self in class functions as much as possible because:
1. We loosely couple those functions with the self params so that it is easier to manage code and possible bugs. I got many troubles and confuses by using self params.
2. Many functions actually become staticmethod by using this way.
3. Code can be reused easily later if some functions are useful.

Reply:
Ok, I'm not as much against using self, especially if it's a method specific to a certain attack and accessing a unique property of this attack which I think is often the case. But could you please mark static methods with @staticmethod if there are any?

Reply:
I have changed to use self params at most of the places. That will help clean most of redundant documents.